### ISSUE TITLE
Support for AssumeRoleWithWebIdentity for S3 module sources

### ISSUE THREAD

**Issue by tiagomeireles** [hashicorp/terraform#27192]
### Current Terraform Version

```
Terraform v0.14.0
```

### Use-cases

Retrieve module archive from S3 while using AWS EKS IAM role for service accounts(IRSA).

### Attempted Solutions

Attempted on Terraform versions 0.12.26, 0.13.5 and 0.14.0. All produce similar errors.

From 0.14.0
```
Error: Failed to download module
Could not download module "foo" (main.tf:61) source code from
"s3::https://s3.amazonaws.com/foo/bar.zip":
NoCredentialProviders: no valid providers in chain
caused by: EnvAccessKeyNotFound: AWS_ACCESS_KEY_ID or AWS_ACCESS_KEY not found
in environment
SharedCredsLoad: failed to load shared credentials file
caused by: FailedRead: unable to open file
caused by: open /root/.aws/credentials: no such file or directory
EC2RoleRequestError: no EC2 instance role found
caused by: EC2MetadataError: failed to make EC2Metadata request
	status code: 404, request id: 
```

### Proposal

Add support for IRSA for s3 module sources.

### References

https://www.terraform.io/docs/modules/sources.html#s3-bucket
https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts-technical-overview.html

**Comment by tiagomeireles** [comment]
@jbardin Is this actually a provider issue? I was under the impression that this functionally was in the core. https://github.com/hashicorp/terraform/blob/6f4f6fbc72360cd8a7a408980b0acd19e8f7f5c1/internal/initwd/getter.go

**Comment by jbardin** [comment]
Thank you @tiagomeireles! I mis-read the request.

I'm not sure this will be supported outside of the normal aws client authentication mechanisms, but we can re-open this here for now.

**Comment by tiagomeireles** [comment]
No worries :). 


This authentication mechanism is supported by other parts of terraform like the aws provider(https://github.com/hashicorp/terraform-provider-aws/pull/14077).

**Comment by dindurthy** [comment]
We're blocked on this too. terraform actions will not detect the service account role and falls back to the instance role or something else. This is a standard AWS authentication mechanism albeit newish.

**Comment by iancward** [comment]
This appears to be a duplicate of #31983, which seems to have been released with 1.5 and backported to 1.4

However, I've tried using both Terraform 1.5 and 1.8 with s3 module sources and Terraform uses the host/EC2 IAM role instead of the Web Identity/IRSA (IAM Role for Service Accounts). But setting `AWS_SDK_LOAD_CONFIG=true` seems to have done the trick.

The [documentation](https://developer.hashicorp.com/terraform/language/modules/sources#s3-bucket) makes no mention of support for IRSA. Can we get the documentation updated to mention that IRSA is supported, but only when `AWS_SDK_LOAD_CONFIG` is set to `true`?
